### PR TITLE
POC: xpu support

### PR DESCRIPTION
**Pull request by sovrasov**
### Summary

- More files are copied from mm*
- ATSSAssigner is temporally replaced with XPUATSSAssigner, which also works in CUDA env

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by harimkang** [comment]
I'm looking at this PR from a v2.0 design perspective.
Overall, it seems like a lot of the dataset, trainer (training api functions) stuff is changing (for mmcv).

We need to support more devices besides XPU.
How about we create something like a `DeviceManager` and collect functions like is_xpu_available, is_cuda_available, is_hpu_available (+...), and functions for supporting device that are agnostic to frameworks? 

(Of course, there will be things that each framework will need to handle depending on the device, but I think the DeviceManager can prevent unnecessary code duplication up front as much as possible. The way it is now, we have to check is_{device}_available with an if statement every time we add a device.)

We have to consider that anomalib is using lightning. Of course, it is expected that lightning already supports many devices, but it would be worth investigating, and it seems like something like this could be done in mmengine (mmcv 2.X).

Actually, for 1.X, the current PR's suggestion might be a good way to go, but I think we need a more general device management design as more frameworks might be added.
**If have any ideas or suggestions, please let me know**, I think I can do a PoC on 2.0 using this code as a reference.

@sovrasov @kprokofi I think your experience can help me a lot. :)

**Comment by sovrasov** [comment]
Now every mm* component has it's own training launcher & model wrapper, so we have to manage devices right in those custom training launchers. In fact, we rely on built-in mm* device management. If this is improved somehow in mmcv 2.0, this can be adjusted on our side as well. Perhaps, a device manager could be created, but first we need to review all the cases including HPU.

**Comment by sovrasov** [comment]
For further considerations it's wroth mentioning that device support consists of 3 basic components:
1. We need to move all the input tensors to the target device -> it's done by wrapping the target model
2. We need to move model's weights to the target device
3. We may need to call a special operator to initialize device-specific optimizer patching (usually that operator requires model+optimizer)

**Comment by eunwoosh** [comment]
I think all of us agree with that we need to make better design. Because Harim already experienced using mmcv 2.0, after we implement HPU in some degree, it's better to discuss about that together.

**Review by eunwoosh — APPROVED** [review]

File: docs/source/guide/explanation/algorithms/visual_prompting/index.rst
Comment (kprokofi) [inline docs/source/guide/explanation/algorithms/visual_prompting/index.rst:1]: Can you update target branch and this one for latest develop? I think it is not related to XPU implementation as well as some other changes

**Comment by kprokofi** [comment]
> I think all of us agree with that we need to make better design. Because Harim already experienced using mmcv 2.0, after we implement HPU in some degree, it's better to discuss about that together.

I guess it is more like PoC -> let's not focus on architecture and code design on this step

File: docs/source/guide/explanation/algorithms/visual_prompting/index.rst
Comment (sovrasov) [reply docs/source/guide/explanation/algorithms/visual_prompting/index.rst:1]: Done

File: src/otx/algorithms/classification/configs/deit_tiny/model.py [hunk 1]
```diff
@@ -14,7 +14,7 @@
     backbone=dict(arch="deit-tiny", init_cfg=dict(type="Pretrained", checkpoint=ckpt_url, prefix="backbone")),
 )
 
-fp16 = dict(loss_scale=512.0)
+# fp16 = dict(loss_scale=512.0)
```
Comment (kprokofi) [inline src/otx/algorithms/classification/configs/deit_tiny/model.py]: Let's make this change in train function, not in templates. Depending on the given device we can set fp16 in config to None (if it is XPU devices).

Comment (sovrasov) [reply src/otx/algorithms/classification/configs/deit_tiny/model.py]: Done

**Review by kprokofi — APPROVED** [review]
LGTM, if tests pass

**Pull request merged**